internal/cache/cache.go

```diff
@@ -0,0 +1,121 @@
+package cache
+
+import (
+	"sync"
+
+	"github.com/rudderlabs/rudder-control-plane-sdk/modelv2"
+	"github.com/rudderlabs/rudder-control-plane-sdk/notifications"
+)
+
+type WorkspaceConfigCache struct {
+	configs        *modelv2.WorkspaceConfigs
+	updateLock     sync.Mutex
+	subscribers    []*Subscriber
+	subscriberLock sync.Mutex
+}
+
+type WorkspaceConfigNotification struct{}
+
+// Get returns a copy of the current workspace configs.
+// This is in order to avoid race conditions when consumers read the configs while they are being updated.
+func (c *WorkspaceConfigCache) Get() *modelv2.WorkspaceConfigs {
+	c.updateLock.Lock()
+	defer c.updateLock.Unlock()
+
+	if c.configs == nil {
+		return nil
+	}
+
+	return copyConfigs(c.configs)
+}
+
+// Set updates the current workspace configs by merging input with current cache contents.
+// Workspace configs are merged by id, so if a workspace config is updated, it will replace the previous one.
+// If a workspace config is not included in input, it will be removed from the cache.
+// If a workspace config is nil in input, it will not be updated.
+// Source and destination definitions are merged without removing any missing definitions.
+// It notifies all subscribers of the update.
+func (c *WorkspaceConfigCache) Set(configs *modelv2.WorkspaceConfigs) {
+	c.updateLock.Lock()
+	c.merge(configs)
+	c.updateLock.Unlock()
+
+	c.subscriberLock.Lock()
+	defer c.subscriberLock.Unlock()
```

Comment:
We can use read locks here, since we are just reading from the slice.
```suggestion
	c.subscriberLock.RLock()
	defer c.subscriberLock.RUnlock()
```